cms/private_media/tests/test_images.py

```diff
@@ -0,0 +1,369 @@
+        - Verify that privacy change tracking works correctly.
+        - Graceful handling of unsupported storage backends.
+        """
+        # Attempts to set file permissions on save should have failed gracefully,
+        # since the default file backend doesn't support it
+        with self.assertLogs("cms.private_media.bulk_operations", level="INFO") as logs:
+            image = ImageFactory(_privacy=Privacy.PUBLIC)
+
+        # This image should be 'public'
+        self.assertIs(image.privacy, Privacy.PUBLIC)
+        self.assertTrue(image.is_public)
+        self.assertFalse(image.is_private)
+
+        # Attempts to set file permissions on save should have failed gracefully
+        self.assertEqual(
+            logs.output,
+            [
+                (
+                    "INFO:cms.private_media.bulk_operations:FileSystemStorage does not support setting of individual "
+                    f"file permissions to public, so skipping for: {image.file.name}."
+                )
+            ],
+        )
+
+        # File permissions should be considered up-to-date
+        self.assertFalse(image.file_permissions_are_outdated())
+
+        # Setting privacy to the same value should not trigger an update to 'privacy_last_changed
+        value_before = image.privacy_last_changed
+        image.privacy = Privacy.PUBLIC
+        self.assertEqual(image.privacy_last_changed, value_before)
+
+        # Setting privacy to a different value should triggered an update to 'privacy_last_changed'
+        image.privacy = Privacy.PRIVATE
+        self.assertGreater(image.privacy_last_changed, value_before)
+
+        # File permissions should now be considered outdated
+        self.assertTrue(image.file_permissions_are_outdated())
+
+        # Resaving should trigger an update to file permissions and the 'file_permissions_last_set'
+        # timestamp, resolving the issue
+        image.save()
+        self.assertFalse(image.file_permissions_are_outdated())
+
+    def test_public_image_renditions(self):
+        """Test rendition.url behaviour for public image renditions:
+        - Public images with up-to-date permissions use direct file URLs.
+        - Falls back to serve URLs when permissions are outdated.
+        """
+        image = ImageFactory(_privacy=Privacy.PUBLIC, collection=self.root_collection)
+        renditions = image.create_renditions(Filter("fill-10x10"), Filter("fill-20x20"))
+        self.assertEqual(len(renditions), 2)
+
+        # For public images with up-to-date file permissions, rendition.url should return the file url
+        for rendition in renditions.values():
+            self.assertEqual(rendition.url, rendition.file.url)
+            # And cache keys should include 'public'
+            self.assertIn("-public-", rendition.get_cache_key())
+
+        # However, if the file permissions are outdated, rendition.url should return a serve URL
```

Comment:
Nit: a lot of the tests in this PR are quite large, and test multiple things in a single case. It might be a good idea to separate them for clarity

Reply:
If they were all `SimpleTestCase` instances, I would agree with you. But, the overhead of doing that for a `TestCase` isn't worth it in my opinion. These tests are well commented and walk through each scenario thoroughly. Things could become more difficult to reason about if broken down further.

Reply:
@RealOrangeOne Happy to mark this one as resolved?